cfgrammar/src/lib/header.rs

```diff
@@ -345,6 +347,132 @@ impl<'input> GrmtoolsSectionParser<'input> {
     }
 }
 
+#[derive(Debug, Eq, PartialEq)]
+pub struct Header {
+    contents: MarkMap<String, (Location, Value)>,
+}
+
+impl Header {
+    pub fn new() -> Self {
+        Header {
+            contents: MarkMap::new(),
+        }
+    }
+
+    pub fn merge_from(
+        &mut self,
+        other: Self,
+    ) -> Result<(), MergeError<String, Box<(Location, Value)>>> {
+        self.contents.merge_from(other.contents)
+    }
+
+    pub fn contents(&self) -> &MarkMap<String, (Location, Value)> {
+        &self.contents
+    }
+
+    pub fn contents_mut(&mut self) -> &mut MarkMap<String, (Location, Value)> {
+        &mut self.contents
+    }
+
+    /// A wrapper around `HashMap::get`, which checks the `query_mask`
+    /// against the `value.query_bits()`
+    ///
+    /// Regardless of whether this query mask check succeeds or fails,
+    /// if the `contents` contains the key, will mark the key as being used.
+    pub fn query(
+        &mut self,
+        key: &str,
+        query_mask: u16,
+    ) -> Option<Result<(&Location, &Value), HeaderError>> {
+        self.contents_mut().mark_used(&key.to_owned());
+        if let Some((key_loc, value)) = self.contents.get(key) {
```

Comment:
I wonder if clippy will complain about this not being `self.contents.get(key).map(...)`?

Reply:
It didn't for me, I don't know why it wouldn't though like maybe option is a special case?

I do kind of dislike `Option::map` and `Result::map`, because I can't tell you how many times i've gotten mixed up by it when I really intended to map over `Some(iterator)`.